File: services/integration/model_mapper.js
/**
 * Model Mapper
 * 
 * Maps data models between Canvas and Discourse systems,
 * providing bidirectional transformation and persistent mapping records.
 */

const { logger } = require('../../shared/logger');
const { db } = require('../../shared/db');

class ModelMapper {
  /**
   * Create a new model mapper instance
   */
  constructor() {
    this.mappingCache = {};
  }

  /**
   * Save a mapping between Canvas and Discourse entities
   * 
   * @param {string} entityType - Type of entity (user, course, etc.)
   * @param {string} sourceId - ID in the source system (usually Canvas)
   * @param {string} targetId - ID in the target system (usually Discourse)
   * @param {string} sourceSystem - Source system ('canvas' or 'discourse')
   * @returns {Object} - Created mapping
   */
  async saveMapping(entityType, sourceId, targetId, sourceSystem = 'canvas') {
    try {
      const targetSystem = sourceSystem === 'canvas' ? 'discourse' : 'canvas';
      
      // Check if mapping already exists
      const existingMapping = await db.entityMappings.findOne({
        where: {
          entityType,
          sourceId,
          sourceSystem
        }
      });
      
      if (existingMapping) {
        // Update existing mapping
        await db.entityMappings.update(
          { targetId },
          {
            where: {
              entityType,
              sourceId,
              sourceSystem
            }
          }
        );
        
        // Update cache
        const cacheKey = `${entityType}:${sourceId}:${sourceSystem}`;
        this.mappingCache[cacheKey] = {
          entityType,
          sourceId,
          sourceSystem,
          targetSystem,
          targetId
        };
        
        return this.mappingCache[cacheKey];
      } else {
        // Create new mapping
        const mapping = await db.entityMappings.create({
          entityType,
          sourceId,
          sourceSystem,
          targetSystem,
          targetId
        });
        
        // Update cache
        const cacheKey = `${entityType}:${sourceId}:${sourceSystem}`;
        this.mappingCache[cacheKey] = mapping;
        
        return mapping;
      }
    } catch (error) {
      logger.error(`Failed to save entity mapping: ${error.message}`);
      throw error;
    }
  }

  /**
   * Get a mapping between Canvas and Discourse entities
   * 
   * @param {string} entityType - Type of entity (user, course, etc.)
   * @param {string} sourceId - ID in the source system
   * @param {string} sourceSystem - Source system ('canvas' or 'discourse')
   * @returns {Object} - Retrieved mapping
   */
  async getMapping(entityType, sourceId, sourceSystem = 'canvas') {
    try {
      // Check cache first
      const cacheKey = `${entityType}:${sourceId}:${sourceSystem}`;
      
      if (this.mappingCache[cacheKey]) {
        return this.mappingCache[cacheKey];
      }
      
      // Query database
      const mapping = await db.entityMappings.findOne({
        where: {
          entityType,
          sourceId,
          sourceSystem
        }
      });
      
      // Update cache if found
      if (mapping) {
        this.mappingCache[cacheKey] = mapping;
      }
      
      return mapping;
    } catch (error) {
      logger.error(`Failed to get entity mapping: ${error.message}`);
      throw error;
    }
  }

  /**
   * Delete a mapping between Canvas and Discourse entities
   * 
   * @param {string} entityType - Type of entity (user, course, etc.)
   * @param {string} sourceId - ID in the source system
   * @param {string} sourceSystem - Source system ('canvas' or 'discourse')
   * @returns {boolean} - Success indicator
   */
  async deleteMapping(entityType, sourceId, sourceSystem = 'canvas') {
    try {
      // Delete from database
      await db.entityMappings.destroy({
        where: {
          entityType,
          sourceId,
          sourceSystem
        }
      });
      
      // Remove from cache
      const cacheKey = `${entityType}:${sourceId}:${sourceSystem}`;
      delete this.mappingCache[cacheKey];
      
      return true;
    } catch (error) {
      logger.error(`Failed to delete entity mapping: ${error.message}`);
      throw error;
    }
  }

  /**
   * Get all mappings for a specific entity type
   * 
   * @param {string} entityType - Type of entity (user, course, etc.)
   * @returns {Array} - List of mappings
   */
  async getAllMappings(entityType) {
    try {
      return await db.entityMappings.findAll({
        where: { entityType }
      });
    } catch (error) {
      logger.error(`Failed to get all mappings: ${error.message}`);
      throw error;
    }
  }

  /**
   * Transform a Canvas user to Discourse user format
   * 
   * @param {Object} canvasUser - Canvas user object
   * @returns {Object} - Discourse user object
   */
  canvasToDiscourseUser(canvasUser) {
    const username = canvasUser.email 
      ? canvasUser.email.split('@')[0].replace(/[^a-zA-Z0-9_]/g, '_')
      : `canvas_user_${canvasUser.id}`;
    
    return {
      name: canvasUser.name,
      username: username,
      email: canvasUser.email,
      password: `canvas_${Date.now()}`, // Will be reset by SSO
      active: canvasUser.workflow_state !== 'deleted',
      approved: true,
      custom_fields: {
        canvas_user_id: canvasUser.id.toString()
      }
    };
  }

  /**
   * Transform a Discourse user to Canvas user format
   * 
   * @param {Object} discourseUser - Discourse user object
   * @returns {Object} - Canvas user object
   */
  discourseToCanvasUser(discourseUser) {
    return {
      name: discourseUser.name,
      short_name: discourseUser.username,
      sortable_name: discourseUser.name,
      email: discourseUser.email,
      status: discourseUser.active ? 'active' : 'deleted'
    };
  }

  /**
   * Transform a Canvas course to Discourse category format
   * 
   * @param {Object} canvasCourse - Canvas course object
   * @returns {Object} - Discourse category object
   */
  canvasToDiscourseCategory(canvasCourse) {
    return {
      name: canvasCourse.name,
      slug: this.slugify(canvasCourse.course_code || canvasCourse.name),
      color: "0088CC", // Default color
      text_color: "FFFFFF",
      description: canvasCourse.public_description || canvasCourse.syllabus_body || '',
      custom_fields: {
        canvas_course_id: canvasCourse.id.toString(),
        canvas_course_code: canvasCourse.course_code
      }
    };
  }

  /**
   * Transform a Discourse category to Canvas course format
   * 
   * @param {Object} discourseCategory - Discourse category object
   * @returns {Object} - Canvas course object
   */
  discourseToCanvasCourse(discourseCategory) {
    const courseCode = discourseCategory.custom_fields?.canvas_course_code || discourseCategory.slug;
    
    return {
      name: discourseCategory.name,
      course_code: courseCode,
      public_description: discourseCategory.description,
      default_view: 'modules'
    };
  }

  /**
   * Transform Canvas assignment to Discourse topic format
   * 
   * @param {Object} canvasAssignment - Canvas assignment object
   * @param {string} categoryId - Discourse category ID
   * @returns {Object} - Discourse topic object
   */
  canvasToDiscourseTopic(canvasAssignment, categoryId) {
    return {
      title: canvasAssignment.name,
      raw: this.formatAssignmentBody(canvasAssignment),
      category: categoryId,
      tags: ['canvas-assignment'],
      custom_fields: {
        canvas_assignment_id: canvasAssignment.id.toString(),
        canvas_points_possible: canvasAssignment.points_possible,
        canvas_due_at: canvasAssignment.due_at
      }
    };
  }

  /**
   * Format assignment description body for Discourse
   * 
   * @param {Object} assignment - Canvas assignment
   * @returns {string} - Formatted content
   */
  formatAssignmentBody(assignment) {
    const parts = [];
    
    parts.push(`# ${assignment.name}`);
    parts.push('');
    
    if (assignment.description) {
      parts.push(assignment.description);
      parts.push('');
    }
    
    parts.push('## Assignment Details');
    parts.push('');
    parts.push(`**Points:** ${assignment.points_possible}`);
    
    if (assignment.due_at) {
      parts.push(`**Due Date:** ${new Date(assignment.due_at).toLocaleString()}`);
    }
    
    if (assignment.submission_types) {
      parts.push(`**Submission Type:** ${assignment.submission_types.join(', ')}`);
    }
    
    parts.push('');
    parts.push(`[View in Canvas](${assignment.html_url})`);
    
    return parts.join('\n');
  }

  /**
   * Transform a Canvas submission to Discourse post format
   * 
   * @param {Object} submission - Canvas submission object
   * @param {string} topicId - Discourse topic ID (for the assignment)
   * @param {string} userId - Discourse user ID
   * @returns {Object} - Discourse post object
   */
  canvasToDiscoursePost(submission, topicId, userId) {
    const content = [];
    
    content.push('**Assignment Submission**');
    content.push('');
    
    if (submission.body) {
      content.push(submission.body);
      content.push('');
    }
    
    if (submission.submission_type === 'online_url' && submission.url) {
      content.push(`Submitted URL: ${submission.url}`);
      content.push('');
    }
    
    if (submission.attachments && submission.attachments.length > 0) {
      content.push('**Attachments:**');
      submission.attachments.forEach(attachment => {
        content.push(`- [${attachment.display_name}](${attachment.url})`);
      });
      content.push('');
    }
    
    content.push(`Submitted at: ${new Date(submission.submitted_at).toLocaleString()}`);
    
    return {
      topic_id: topicId,
      user_id: userId, 
      raw: content.join('\n'),
      custom_fields: {
        canvas_submission_id: submission.id.toString(),
        canvas_submission_type: submission.submission_type,
        canvas_submitted_at: submission.submitted_at
      }
    };
  }

  /**
   * Transform Canvas grade data to Discourse post update
   * 
   * @param {Object} submission - Canvas submission with grade
   * @param {string} postId - Discourse post ID
   * @returns {Object} - Discourse post update object
   */
  canvasGradeToDiscoursePost(submission, postId) {
    const content = [];
    
    // Original post content is preserved by Discourse API
    // We're adding a reply from the instructor
    content.push('**Grade Information**');
    content.push('');
    
    if (submission.grade) {
      content.push(`**Score:** ${submission.grade}/${submission.assignment.points_possible}`);
    }
    
    if (submission.score) {
      content.push(`**Points:** ${submission.score}`);
    }
    
    if (submission.graded_at) {
      content.push(`**Graded:** ${new Date(submission.graded_at).toLocaleString()}`);
    }
    
    if (submission.grader_id) {
      content.push(`**Grader ID:** ${submission.grader_id}`);
    }
    
    if (submission.comment) {
      content.push('');
      content.push('**Instructor Comment:**');
      content.push(submission.comment);
    }
    
    return {
      id: postId,
      raw: content.join('\n'),
      custom_fields: {
        canvas_grade: submission.grade,
        canvas_score: submission.score,
        canvas_graded_at: submission.graded_at
      }
    };
  }

  /**
   * Transform Canvas discussion to Discourse topic
   * 
   * @param {Object} discussion - Canvas discussion object
   * @param {string} categoryId - Discourse category ID
   * @returns {Object} - Discourse topic object
   */
  canvasToDiscourseDiscussion(discussion, categoryId) {
    return {
      title: discussion.title,
      raw: discussion.message || discussion.title,
      category: categoryId,
      tags: ['canvas-discussion'],
      custom_fields: {
        canvas_discussion_id: discussion.id.toString(),
        canvas_discussion_type: discussion.discussion_type
      }
    };
  }

  /**
   * Slugify a string for use in URLs
   * 
   * @param {string} text - Text to slugify
   * @returns {string} - Slugified text
   */
  slugify(text) {
    return text
      .toString()
      .toLowerCase()
      .replace(/\s+/g, '-')
      .replace(/[^\w\-]+/g, '')
      .replace(/\-\-+/g, '-')
      .replace(/^-+/, '')
      .replace(/-+$/, '');
  }
}

module.exports = {
  ModelMapper
};

File: services/integration/model_mapper.rs
use crate::shared::logger::Logger;
use crate::shared::db::Database;
use serde::{Serialize, Deserialize};
use std::sync::Arc;
use std::collections::HashMap;
use tokio::sync::Mutex;
use thiserror::Error;

/// Error types for the model mapper
#[derive(Error, Debug)]
pub enum MapperError {
    #[error("Database error: {0}")]
    DatabaseError(String),
    
    #[error("Invalid entity type: {0}")]
    InvalidEntityType(String),
    
    #[error("Invalid system name: {0}")]
    InvalidSystemName(String),
    
    #[error("Mapping not found: {0}/{1}")]
    MappingNotFound(String, String),
    
    #[error("Caching error: {0}")]
    CachingError(String),
}

/// Entity mapping structure
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EntityMapping {
    pub entity_type: String,
    pub source_id: String,
    pub source_system: String,
    pub target_system: String,
    pub target_id: String,
}

/// Cache key for entity mappings
type CacheKey = String;

/// Model Mapper
/// 
/// Maps data models between Canvas and Discourse systems,
/// providing bidirectional transformation and persistent mapping records.
pub struct ModelMapper {
    logger: Arc<Logger>,
    db: Arc<Database>,
    mapping_cache: Mutex<HashMap<CacheKey, EntityMapping>>,
}

impl ModelMapper {
    /// Create a new model mapper instance
    pub fn new(logger: Arc<Logger>, db: Arc<Database>) -> Self {
        ModelMapper {
            logger,
            db,
            mapping_cache: Mutex::new(HashMap::new()),
        }
    }
    
    /// Save a mapping between Canvas and Discourse entities
    /// 
    /// # Arguments
    /// 
    /// * `entity_type` - Type of entity (user, course, etc.)
    /// * `source_id` - ID in the source system (usually Canvas)
    /// * `target_id` - ID in the target system (usually Discourse)
    /// * `source_system` - Source system ('canvas' or 'discourse')
    /// 
    /// # Returns
    /// 
    /// * `Result<EntityMapping, MapperError>` - Created or updated mapping
    pub async fn save_mapping(
        &self,
        entity_type: &str,
        source_id: &str,
        target_id: &str,
        source_system: Option<&str>,
    ) -> Result<EntityMapping, MapperError> {
        let source_system = source_system.unwrap_or("canvas");
        self.validate_system_name(source_system)?;
        
        // Determine target system
        let target_system = if source_system == "canvas" { "discourse" } else { "canvas" };
        
        // Create cache key
        let cache_key = format!("{}:{}:{}", entity_type, source_id, source_system);
        
        // Check if mapping already exists
        let existing_mapping = self.db.find_entity_mapping(entity_type, source_id, source_system)
            .await
            .map_err(|e| MapperError::DatabaseError(e.to_string()))?;
        
        let mapping = if let Some(existing) = existing_mapping {
            // Update existing mapping
            self.db.update_entity_mapping(entity_type, source_id, source_system, target_id)
                .await
                .map_err(|e| MapperError::DatabaseError(e.to_string()))?;
            
            EntityMapping {
                entity_type: entity_type.to_string(),
                source_id: source_id.to_string(),
                source_system: source_system.to_string(),
                target_system: target_system.to_string(),
                target_id: target_id.to_string(),
            }
        } else {
            // Create new mapping
            self.db.create_entity_mapping(entity_type, source_id, source_system, target_system, target_id)
                .await
                .map_err(|e| MapperError::DatabaseError(e.to_string()))?
        };
        
        // Update cache
        let mut cache = self.mapping_cache.lock().await;
        cache.insert(cache_key, mapping.clone());
        
        Ok(mapping)
    }
    
    /// Get a mapping between Canvas and Discourse entities
    /// 
    /// # Arguments
    /// 
    /// * `entity_type` - Type of entity (user, course, etc.)
    /// * `source_id` - ID in the source system
    /// * `source_system` - Source system ('canvas' or 'discourse')
    /// 
    /// # Returns
    /// 
    /// * `Result<EntityMapping, MapperError>` - Retrieved mapping
    pub async fn get_mapping(
        &self,
        entity_type: &str,
        source_id: &str,
        source_system: Option<&str>,
    ) -> Result<EntityMapping, MapperError> {
        let source_system = source_system.unwrap_or("canvas");
        self.validate_system_name(source_system)?;
        
        // Create cache key
        let cache_key = format!("{}:{}:{}", entity_type, source_id, source_system);
        
        // Check cache first
        {
            let cache = self.mapping_cache.lock().await;
            if let Some(mapping) = cache.get(&cache_key) {
                return Ok(mapping.clone());
            }
        }
        
        // Cache miss - query the database
        let mapping = self.db.find_entity_mapping(entity_type, source_id, source_system)
            .await
            .map_err(|e| MapperError::DatabaseError(e.to_string()))?
            .ok_or_else(|| MapperError::MappingNotFound(entity_type.to_string(), source_id.to_string()))?;
        
        // Update cache
        let mut cache = self.mapping_cache.lock().await;
        cache.insert(cache_key, mapping.clone());
        
        Ok(mapping)
    }
    
    /// Delete a mapping between Canvas and Discourse entities
    /// 
    /// # Arguments
    /// 
    /// * `entity_type` - Type of entity (user, course, etc.)
    /// * `source_id` - ID in the source system
    /// * `source_system` - Source system ('canvas' or 'discourse')
    /// 
    /// # Returns
    /// 
    /// * `Result<bool, MapperError>` - True if mapping was deleted
    pub async fn delete_mapping(
        &self,
        entity_type: &str,
        source_id: &str,
        source_system: Option<&str>,
    ) -> Result<bool, MapperError> {
        let source_system = source_system.unwrap_or("canvas");
        self.validate_system_name(source_system)?;
        
        // Create cache key
        let cache_key = format!("{}:{}:{}", entity_type, source_id, source_system);
        
        // Delete from database
        let deleted = self.db.delete_entity_mapping(entity_type, source_id, source_system)
            .await
            .map_err(|e| MapperError::DatabaseError(e.to_string()))?;
        
        // Remove from cache if exists
        if deleted {
            let mut cache = self.mapping_cache.lock().await;
            cache.remove(&cache_key);
        }
        
        Ok(deleted)
    }
    
    /// Get the target ID for a source entity
    /// 
    /// # Arguments
    /// 
    /// * `entity_type` - Type of entity (user, course, etc.)
    /// * `source_id` - ID in the source system
    /// * `source_system` - Source system ('canvas' or 'discourse')
    /// 
    /// # Returns
    /// 
    /// * `Result<String, MapperError>` - Target ID
    pub async fn get_target_id(
        &self,
        entity_type: &str,
        source_id: &str,
        source_system: Option<&str>,
    ) -> Result<String, MapperError> {
        let mapping = self.get_mapping(entity_type, source_id, source_system).await?;
        Ok(mapping.target_id)
    }
    
    /// Find all mappings for an entity type
    /// 
    /// # Arguments
    /// 
    /// * `entity_type` - Type of entity (user, course, etc.)
    /// 
    /// # Returns
    /// 
    /// * `Result<Vec<EntityMapping>, MapperError>` - List of entity mappings
    pub async fn find_mappings_by_type(
        &self,
        entity_type: &str,
    ) -> Result<Vec<EntityMapping>, MapperError> {
        self.db.find_entity_mappings_by_type(entity_type)
            .await
            .map_err(|e| MapperError::DatabaseError(e.to_string()))
    }
    
    /// Clear the mapping cache
    pub async fn clear_cache(&self) {
        let mut cache = self.mapping_cache.lock().await;
        cache.clear();
    }
    
    /// Helper method to validate system names
    fn validate_system_name(&self, system_name: &str) -> Result<(), MapperError> {
        match system_name {
            "canvas" | "discourse" => Ok(()),
            _ => Err(MapperError::InvalidSystemName(system_name.to_string())),
        }
    }
}
